File: repos/safewords-ios/Safewords/Crypto/RecoveryPhrase.swift
import CryptoKit
import Foundation

enum RecoveryPhraseError: Error, Equatable {
    case emptyInput
    case wrongWordCount
    case unknownWord(index: Int, word: String)
    case badChecksum

    var code: String {
        switch self {
        case .emptyInput:
            return "EMPTY_INPUT"
        case .wrongWordCount:
            return "WRONG_WORD_COUNT"
        case .unknownWord:
            return "UNKNOWN_WORD"
        case .badChecksum:
            return "BAD_CHECKSUM"
        }
    }

    var userMessage: String {
        switch self {
        case .emptyInput:
            return "Enter your 24-word recovery phrase."
        case .wrongWordCount:
            return "Recovery phrase must be exactly 24 words."
        case let .unknownWord(index, word):
            return "Word \(index) is not in the recovery word list: \"\(word)\"."
        case .badChecksum:
            return "Recovery phrase checksum is invalid. Check the words and order."
        }
    }
}

extension RecoveryPhraseError: LocalizedError {
    var errorDescription: String? {
        userMessage
    }
}

struct Bip39 {
    static let expectedWordCount = 24
    static let expectedEntropyBytes = 32
    static let expectedVocabularySize = 2_048

    private static let checksumBits = 8
    private static let bitsPerWord = 11

    let wordlist: [String]
    private let wordIndex: [String: Int]

    init(wordlist: [String]) {
        precondition(
            wordlist.count == Self.expectedVocabularySize,
            "BIP39 English wordlist must have \(Self.expectedVocabularySize) words (got \(wordlist.count))"
        )
        self.wordlist = wordlist
        self.wordIndex = Dictionary(uniqueKeysWithValues: wordlist.enumerated().map { ($0.element, $0.offset) })
    }

    func encode(seed: Data) throws -> String {
        guard seed.count == Self.expectedEntropyBytes else {
            throw RecoveryPhraseRuntimeError.invalidSeedLength(seed.count)
        }

        let checksumByte = Self.sha256(seed)[0]
        var indices: [Int] = []
        indices.reserveCapacity(Self.expectedWordCount)

        var bitBuffer: UInt64 = 0
        var bitsInBuffer = 0

        for byte in seed {
            bitBuffer = (bitBuffer << 8) | UInt64(byte)
            bitsInBuffer += 8
            drainWords(from: &bitBuffer, bitsInBuffer: &bitsInBuffer, into: &indices)
        }

        bitBuffer = (bitBuffer << Self.checksumBits) | UInt64(checksumByte)
        bitsInBuffer += Self.checksumBits
        drainWords(from: &bitBuffer, bitsInBuffer: &bitsInBuffer, into: &indices)

        guard indices.count == Self.expectedWordCount else {
            throw RecoveryPhraseRuntimeError.invalidWordIndexCount(indices.count)
        }

        return indices.map { wordlist[$0] }.joined(separator: " ")
    }

    func decode(input: String) throws -> Data {
        let tokens = normalize(input: input)
        if tokens.isEmpty {
            throw RecoveryPhraseError.emptyInput
        }
        guard tokens.count == Self.expectedWordCount else {
            throw RecoveryPhraseError.wrongWordCount
        }

        let indices = try tokens.enumerated().map { offset, token in
            guard let index = wordIndex[token] else {
                throw RecoveryPhraseError.unknownWord(index: offset + 1, word: token)
            }
            return index
        }

        var seed = Data()
        seed.reserveCapacity(Self.expectedEntropyBytes)
        var bitBuffer: UInt64 = 0
        var bitsInBuffer = 0

        for index in indices {
            bitBuffer = (bitBuffer << Self.bitsPerWord) | UInt64(index)
            bitsInBuffer += Self.bitsPerWord
            while bitsInBuffer >= 8 && seed.count < Self.expectedEntropyBytes {
                let shift = bitsInBuffer - 8
                seed.append(UInt8((bitBuffer >> shift) & 0xff))
                bitsInBuffer -= 8
                bitBuffer &= Self.lowBitMask(bitsInBuffer)
            }
        }

        guard bitsInBuffer == Self.checksumBits else {
            throw RecoveryPhraseRuntimeError.invalidTrailingBitCount(bitsInBuffer)
        }

        let providedChecksum = UInt8(bitBuffer & 0xff)
        let expectedChecksum = Self.sha256(seed)[0]
        guard providedChecksum == expectedChecksum else {
            throw RecoveryPhraseError.badChecksum
        }

        return seed
    }

    func normalize(input: String) -> [String] {
        let normalized = input.decomposedStringWithCompatibilityMapping
        let lowered = normalized
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased(with: Locale(identifier: "en_US_POSIX"))

        guard !lowered.isEmpty else { return [] }
        return lowered.split(whereSeparator: \.isWhitespace).map(String.init)
    }

    private func drainWords(from bitBuffer: inout UInt64, bitsInBuffer: inout Int, into indices: inout [Int]) {
        while bitsInBuffer >= Self.bitsPerWord {
            let shift = bitsInBuffer - Self.bitsPerWord
            indices.append(Int((bitBuffer >> shift) & 0x7ff))
            bitsInBuffer -= Self.bitsPerWord
            bitBuffer &= Self.lowBitMask(bitsInBuffer)
        }
    }

    private static func sha256(_ data: Data) -> [UInt8] {
        Array(SHA256.hash(data: data))
    }

    private static func lowBitMask(_ bits: Int) -> UInt64 {
        bits == 0 ? 0 : (UInt64(1) << bits) - 1
    }
}

enum RecoveryPhrase {
    private static let bip39: Bip39 = {
        Bip39(wordlist: loadWordlist(from: .main))
    }()

    static func encode(seed: Data) throws -> String {
        try bip39.encode(seed: seed)
    }

    static func decode(input: String) throws -> Data {
        try bip39.decode(input: input)
    }

    static func normalize(input: String) -> [String] {
        bip39.normalize(input: input)
    }

    static func loadWordlist(from bundle: Bundle) -> [String] {
        guard let url = bundle.url(forResource: "bip39-english", withExtension: "txt"),
              let raw = try? String(contentsOf: url, encoding: .utf8) else {
            fatalError("bip39-english.txt is missing from bundle resources")
        }
        return raw
            .split(separator: "\n")
            .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
            .filter { !$0.isEmpty }
    }
}

private enum RecoveryPhraseRuntimeError: Error {
    case invalidSeedLength(Int)
    case invalidWordIndexCount(Int)
    case invalidTrailingBitCount(Int)
}

File: repos/safewords-android/app/src/main/kotlin/com/thc/safewords/crypto/RecoveryPhrase.kt
package com.thc.safewords.crypto

import com.thc.safewords.SafewordsApp
import java.security.MessageDigest
import java.text.Normalizer
import java.util.Locale

/**
 * BIP39 English 24-word recovery phrase encoding for 256-bit (32-byte) Safewords
 * group seeds. Implements the contract at /shared/recovery-schema.md.
 *
 * Entropy-only: this does NOT derive a 64-byte BIP39 seed via PBKDF2/passphrase.
 * Decoding a valid phrase returns the original 32-byte group seed directly.
 */
class Bip39(private val wordlist: List<String>) {

    init {
        if (wordlist.size != EXPECTED_VOCAB_SIZE) {
            throw Error.InvalidWordlistCount(wordlist.size)
        }
    }

    private val wordIndex: Map<String, Int> = wordlist.withIndex().associate { (i, w) -> w to i }

    sealed class Error(val code: String, val userMessage: String) : RuntimeException(userMessage) {
        data object EmptyInput : Error(
            code = "EMPTY_INPUT",
            userMessage = "Enter your 24-word recovery phrase."
        )
        data object WrongWordCount : Error(
            code = "WRONG_WORD_COUNT",
            userMessage = "Recovery phrase must be exactly 24 words."
        )
        class UnknownWord(val wordIndex: Int, val word: String) : Error(
            code = "UNKNOWN_WORD",
            userMessage = "Word $wordIndex is not in the recovery word list: \"$word\"."
        )
        data object BadChecksum : Error(
            code = "BAD_CHECKSUM",
            userMessage = "Recovery phrase checksum is invalid. Check the words and order."
        )
        data object MissingWordlist : Error(
            code = "MISSING_WORDLIST",
            userMessage = "Recovery word list is unavailable. Use the raw seed backup or reinstall the app."
        )
        class InvalidWordlistCount(val count: Int) : Error(
            code = "INVALID_WORDLIST",
            userMessage = "Recovery word list is invalid. Expected 2048 words, found $count."
        )
    }

    /**
     * Encode a 32-byte seed as a 24-word BIP39 English phrase.
     */
    fun encode(seed: ByteArray): String {
        require(seed.size == EXPECTED_ENTROPY_BYTES) {
            "Seed must be $EXPECTED_ENTROPY_BYTES bytes (got ${seed.size})"
        }
        val checksumByte = sha256(seed)[0]
        val indices = IntArray(EXPECTED_WORD_COUNT)
        var bitBuffer = 0L
        var bitsInBuffer = 0
        var indexCursor = 0
        // Push entropy bytes
        for (b in seed) {
            bitBuffer = (bitBuffer shl 8) or (b.toLong() and 0xFF)
            bitsInBuffer += 8
            while (bitsInBuffer >= BITS_PER_WORD) {
                val shift = bitsInBuffer - BITS_PER_WORD
                indices[indexCursor++] = ((bitBuffer ushr shift) and 0x7FF).toInt()
                bitsInBuffer -= BITS_PER_WORD
                bitBuffer = bitBuffer and ((1L shl bitsInBuffer) - 1)
            }
        }
        // Push 8 checksum bits
        bitBuffer = (bitBuffer shl CHECKSUM_BITS) or (checksumByte.toLong() and 0xFF)
        bitsInBuffer += CHECKSUM_BITS
        while (bitsInBuffer >= BITS_PER_WORD) {
            val shift = bitsInBuffer - BITS_PER_WORD
            indices[indexCursor++] = ((bitBuffer ushr shift) and 0x7FF).toInt()
            bitsInBuffer -= BITS_PER_WORD
            bitBuffer = bitBuffer and ((1L shl bitsInBuffer) - 1)
        }
        check(indexCursor == EXPECTED_WORD_COUNT) {
            "Bit-packing produced $indexCursor words, expected $EXPECTED_WORD_COUNT"
        }
        return indices.joinToString(" ") { wordlist[it] }
    }

    /**
     * Decode a 24-word phrase back to its 32-byte seed.
     * Throws [Error] subclasses on invalid input.
     */
    fun decode(input: String): ByteArray {
        val tokens = normalize(input)
        if (tokens.isEmpty()) throw Error.EmptyInput
        if (tokens.size != EXPECTED_WORD_COUNT) throw Error.WrongWordCount

        val indices = IntArray(EXPECTED_WORD_COUNT)
        for ((i, token) in tokens.withIndex()) {
            val idx = wordIndex[token] ?: throw Error.UnknownWord(wordIndex = i + 1, word = token)
            indices[i] = idx
        }

        // Rebuild 264 bits from 24 11-bit indices
        val seed = ByteArray(EXPECTED_ENTROPY_BYTES)
        var bitBuffer = 0L
        var bitsInBuffer = 0
        var byteCursor = 0
        for (idx in indices) {
            bitBuffer = (bitBuffer shl BITS_PER_WORD) or (idx.toLong() and 0x7FF)
            bitsInBuffer += BITS_PER_WORD
            while (bitsInBuffer >= 8 && byteCursor < EXPECTED_ENTROPY_BYTES) {
                val shift = bitsInBuffer - 8
                seed[byteCursor++] = ((bitBuffer ushr shift) and 0xFF).toByte()
                bitsInBuffer -= 8
                bitBuffer = bitBuffer and ((1L shl bitsInBuffer) - 1)
            }
        }
        // Remaining bitsInBuffer bits are the checksum (should be 8)
        check(bitsInBuffer == CHECKSUM_BITS) {
            "Bit-unpacking left $bitsInBuffer trailing bits, expected $CHECKSUM_BITS"
        }
        val providedChecksum = (bitBuffer and 0xFF).toInt()
        val expectedChecksum = sha256(seed)[0].toInt() and 0xFF
        if (providedChecksum != expectedChecksum) throw Error.BadChecksum

        return seed
    }

    /**
     * Apply schema-mandated normalization to user input:
     * NFKD → trim → Locale.ROOT lowercase → split on whitespace.
     */
    fun normalize(input: String): List<String> {
        val nfkd = Normalizer.normalize(input, Normalizer.Form.NFKD)
        val lower = nfkd.trim().lowercase(Locale.ROOT)
        if (lower.isEmpty()) return emptyList()
        return lower.split(Regex("\\s+")).filter { it.isNotEmpty() }
    }

    private fun sha256(input: ByteArray): ByteArray =
        MessageDigest.getInstance("SHA-256").digest(input)

    companion object {
        const val EXPECTED_WORD_COUNT = 24
        const val EXPECTED_ENTROPY_BYTES = 32
        const val EXPECTED_VOCAB_SIZE = 2048
        private const val CHECKSUM_BITS = 8
        private const val BITS_PER_WORD = 11
    }
}

/**
 * App-runtime singleton that loads the BIP39 English wordlist from app assets.
 * For unit tests, construct [Bip39] directly with a wordlist read from disk.
 */
object RecoveryPhrase {
    /**
     * Lazy-loads the wordlist on first use. If the asset is missing or
     * malformed, surfaces the failure as a Bip39.Error so callers can fall
     * back to hex seed display rather than crashing the app.
     */
    private val instanceResult: Result<Bip39> by lazy {
        runCatching {
            val stream = try {
                SafewordsApp.instance.assets.open("wordlists/bip39-english.txt")
            } catch (_: Throwable) {
                throw Bip39.Error.MissingWordlist
            }
            val raw = stream.bufferedReader(Charsets.UTF_8).use { it.readText() }
            val words = raw.split('\n').map { it.trim() }.filter { it.isNotEmpty() }
            Bip39(words)
        }
    }

    private fun instance(): Bip39 = instanceResult.getOrThrow()

    fun encode(seed: ByteArray): String = instance().encode(seed)
    fun decode(input: String): ByteArray = instance().decode(input)
    fun normalize(input: String): List<String> = instance().normalize(input)
}
